### ISSUE TITLE
Add EE capability to Batch TCK to eliminate needing to contribute to Platform TCK  

### ISSUE THREAD

**Issue by scottkurz** [eclipse-ee4j/batch-tck#18]
The TCK situation is rather confusing.   To review: 

* The standalone TCK in this repo (https://github.com/eclipse-ee4j/batch-tck) can be used to verify a Jakarta Batch impl on an SE platform (of course, note Jakarta Batch isn't itself part of the SE platform). 
* This standalone TCK is what we use to certify the current [jbatch](https://github.com/WASdev/standards.jsr352.jbatch) CI.   This is an important, necessary step in the process of releasing updates to the specification/API
* The Jakarta Platform TCK (for batch tests browse to:  https://github.com/eclipse-ee4j/jakartaee-tck/tree/master/src/com/ibm/jbatch/tck/tests) has basically a fork of this same set of Batch tests... plus a few others, (testing transactions in EE only).
* In the past, IBM basically "forked" or dual-contributed these tests both to the ancestor of the open source TCK and the ancestor of the Platform TCK.  
* Presently a batch impl could demonstrate Batch compliance either within the Platform TCK or with our standalone TCK.

From the perspective of Jakarta Batch, I think we want to:

1. Enable the batch-tck to be executed within an EE environment, by "wrapping" the tests somehow
2. Get out of the business of forking and contributing our tests to two places

The downside here is from the perspective of someone who wants to certify their impl against the EE platform (i.e. against batch along with the rest of the EE platform).   This proposal potentially adds a burden on them to setup and configure a whole other TCK, whereas today they are just running Batch using the existing Platform TCK automation.    I understand that each of the specs:   Jakarta Dependency Inection , Contexts and Dependency Inection , Bean Validation and XML Binding  provide a standalone TCK which validates an impl within the platform.      Perhaps we can look into an opportunity to reuse config in some form?

(Note:  besides the runtime tests we also have the signature tests, also "forked" across batch-tck and the platform TCK.   To my knowledge, the running of the signature tests allows for some flexibility in how you set up the "classpath" and isn't overly prescriptive on the difference between running in SE and EE).

**Comment by scottkurz** [comment]
See also #15

**Comment by OndroMih** [comment]
Hi @scottkurz , I propose a solution based on JUnit 5 and Arquillian that I explored last year: https://github.com/OndroMih/batch-tck/pull/3. This solution allows running tests with plain JUnit5 for flat implementations that can run on maven's classpath, and it also allows running the tests on remote containers like Application servers using Arquillian. The tests themselves only depend on JUnit5. The Arquillian support is just an add-on that works as an interceptor for the tests.

Since last year, Arquillian now provides an official JUnit5 extension in Maven Central so everything can be integrated smoothly.

If you like it, I can update my proof of concept so that it's based on the latest commit in the master branch and submit it as a PR. 

The question is whether we're willing to switch from TestNG to JUnit 5. I'm in favor of JUnit 5 for multiple reasons:

* it's more modern than TestNG
* It's more extensible
* It allows intercepting the tests and transfer execution to a remote container without any modification of the test code (TestNG requires that all test classes extend a custom TestNG runner to do that)
* Rewriting the current tests from TestNG to JUnit 5 is rather straightforward and I already did it

However, I can also imagine a solution with Test NG if you like to keep Test NG. But it could be a lot more work to find a way how to run the tests in both modes with either our own Test NG runner or Arquillian.

By the way, if we use Arquillian to run on external runtimes, then it's easy to synchronize with the other standalone TCKs you mentioned because all of them except DI use Arquillian.

**Comment by scottkurz** [comment]
Thanks @OndroMih .  

Re: TestNG, the main thing we took advantage of was the XML-defined "suite".   This gave us a way to define the EE vs SE execution from a single TCK package.   Looks like your PR has a solution.  Also I think Romain has mentioned other TCKs use directory patterns to set up profiles to include/exclude.  

One could note it also gave us a nice way to add "exclusions" in response to TCK issues/challenges, etc., but I don't think that's a strong case.   The hard part of making any update is all the rest of the formal process: tracking, release, etc., rather than compilation itself.   

Let me find some time to try running this against Open Liberty and then we can take this to the list and see what others think.

Having a similar approach as other standalone TCKs would be a strong positive, if not strictly from the batch perspective, then of course because many implementations are interested in certifying the whole platform.

**Comment by OndroMih** [comment]
I agree it would be nice to have the same solution used by all standalone TCKs. But there are now only 4 standalone TCKs and even those do not use the same solution as I described in [my blog](https://ondro.inginea.eu/index.php/possible-ways-to-use-arquillian-in-jakarta-ee-tcks/). 2 use TestNG + Arquillian, 1 uses Junit 4 + Arquillian, 1 uses plain JUnit4. And those that use Arquillian need a special adapter to run tests in plain Java SE.

The solution I propose uses JUnit 5 which is more modern and flexible than JUnit 4 and TestNG. I also proved that migrating tests from Test NG to JUnit 5 is a simple exercise, it took me 1-2 hours to refactor all tests in JBatch.

And it also doesn't require to run tests via Arquillian because JUnit 5 is so flexible that it's possible to write tests in plain JUNit 5 and then add an extension that bundles them in Arquillian deployment and runs them in an Arquillian container.

Moreover, all of the solutions, either existing ones or the new one I propose, run tests using Maven and can be consumed in a unified way by the umbrella TCK.

If you'd like to have a unified approach, I can start the discussion about migrating to JUnit 5 in the other standalone TCKs and also in the Jakarta EE TCK project.

**Comment by OndroMih** [comment]
Meanwhile, I also have a solution to run the existing JBatch tests in Arquillian and thus run the EE tests in containers that support Jakarta EE, here: [poc-arquillian](https://github.com/OndroMih/batch-tck/tree/poc-arquillian). We can start with that to resolve this issue and then discuss moving to JUnit 5 later after a discussion with the other projects.

Would you like me to first submit a PR from my poc-arquillian branch to run EE tests via Arquillian on top of TestNG, @scottkurz ?

**Comment by scottkurz** [comment]
Hi @OndroMih .. just taking a look at your PRs now.

I have a bit of learning curve ahead not being that familiar with Arquillian.

> Would you like me to first submit a PR from my poc-arquillian branch to run EE tests via Arquillian on top of TestNG, @scottkurz 

Would such a PR basically be this:    https://github.com/OndroMih/batch-tck/pull/4 rebased to 'master' ?     

(I was able to merge this last PR trivially.  OTOH, the Junit 5 one,  https://github.com/OndroMih/batch-tck/pull/3, I hit some whitespace issues since we've gotten rid of tabs in the master now.)

 I know you know Romain had some ideas and opinions here.   

I don't have anything more intelligent to say until I understand this a bit better, but will make this a priority.  Thanks for moving the discussion ahead.

**Comment by scottkurz** [comment]
This is already possible via Ondro's work.  It just needs to be documented in the TCK guide. Thx Ondro !
